File: threadring/src/threadring.go
/* Compile and run under Linux and macOS with "go run threadring.go"

Synchronous channels are used for passing the token between the node goroutines.
Each node goroutine continuously accepts tokens; if the token is larger than 0,
it passes the token minus 1 on, otherwise it prints its id.

In Go, a program terminates when the main program terminates. Here, the main
program waits to receive a notification from the channel done; the node that
has the token with value 0 and prints it notifies the main program through
the channel done.
*/
package main
import ("fmt"
        "os"
        "strconv")

// const Hops = 100000
// const Nodes = 1000
func node(n int, in chan int, out chan int, done chan bool) {
  for v := range in {
    if v > 0 {out <- v - 1
    } else {done <- true
    }
  }
}
func main() {
  if len(os.Args) != 3{
    fmt.Printf("Usage: %s Hops Nodes\n", os.Args[0])
    return
  }
  ArgHops := os.Args[1]
  ArgNodes := os.Args[2]
  Hops, errC := strconv.Atoi(ArgHops)
  if errC != nil {
    fmt.Println(errC)
    os.Exit(2)
  }  
  Nodes, errR := strconv.Atoi(ArgNodes)
  if errR != nil {
    fmt.Println(errR)
    os.Exit(2)
  }
  done, ch0 := make(chan bool), make(chan int)
  ch := ch0
  for n := 1; n < Nodes; n++ {
    ch1 := make(chan int)
    go node(n, ch0, ch1, done)
    ch0 = ch1
  }
  go node(0, ch0, ch, done)
	ch0 <- Hops; <- done
}

File: threadring/src/threadring.rs
/* Compile under Linux and macOS with "rustc threadring.rs"; run with "./threadring".

Synchronous channels from the mpsc library (multiple producer, single consumer) are used for
passing the token between the node threads. Each thread continuously accepts tokens; if the token
is larger than 0, it passes the token minus 1 on, otherwise it prints its id. As each node owns the
transmitting and receiving ends of its channels, one transmitting end needs to be cloned for
sending the initial value of the token to a node thread.

Rust requires all threads to terminate properly for the program to terminate properly. For this,
each node thread terminates if it won't receive any more tokens, which is when the received token
is less than the number of nodes.
*/

use std::sync::mpsc::*;
use std::thread::spawn;
use std::env;

// const HOPS: u32 = 100000; // HOPS must be greater than NODES ...
// const NODES: u32 = 1000;  // ... for proper termination

fn node(n: u32, tx: SyncSender<u32>, rx: Receiver<u32>, nodes: u32) {
  loop {
    let v = rx.recv().unwrap();
    if v > 0 {tx.send(v - 1).unwrap()
    } else {//println!("{}", n)
    }
    if v < nodes {break}
  }
}
fn main() {
  let args: Vec<String> = env::args().collect();
  let hops: u32= args[1].parse().unwrap();
  let nodes: u32 = args[2].parse().unwrap();
  let (tx0, mut rx0) = sync_channel(0);
  for n in 1 .. nodes {
    let (tx1, rx1) = sync_channel(0);
    spawn(move || node(n, tx1, rx0, nodes));
    rx0 = rx1
  }
  let tx = SyncSender::clone(&tx0);
  let handle = spawn(move || node(0, tx0, rx0, nodes));
  tx.send(hops).unwrap();
  handle.join().unwrap()
}
